application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java

```diff
@@ -203,11 +219,100 @@ public void updateData(String fromVersion) throws Exception {
                     log.info("Skipping edge events migration");
                 }
                 break;
+            case "3.5.0":
+                log.info("Updating data from version 3.5.0 to 3.5.1 ...");
+                upgradeRuleNodes();
+                break;
             default:
                 throw new RuntimeException("Unable to update data, unsupported fromVersion: " + fromVersion);
         }
     }
 
+    private void upgradeRuleNodes() {
+        var ruleChainIdToTenantIdMap = new HashMap<RuleChainId, TenantId>();
+        try {
+            log.info("Starting rule nodes upgrade ...");
+            var ruleNodeDefinitions = getBeanDefinitions(
+                    org.thingsboard.rule.engine.api.RuleNode.class
+            );
+            for (BeanDefinition def : ruleNodeDefinitions) {
+                String clazzName = def.getBeanClassName();
+                Class<?> clazz = Class.forName(clazzName);
+                TbNode tbNode = (TbNode) clazz.getDeclaredConstructor().newInstance();
+                if (tbNode instanceof VersionedNode) {
```

Comment:
Use VersionedNode.class.isAssignableFrom(clazz). This will help to avoid unnecessary construction of new instances of the node if it is not VersionedNode.
https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html#isAssignableFrom-java.lang.Class-

Reply:
I also believe it is safe to rename ```VersionedNode to TbVersionedNode``` and make ```TbVersionedNode extends TbNode```